GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -153,27 +197,37 @@ public bool TryGetRowIndex(ObjectId objectId, out int index)
                 return false;
             }
 
-            index = Array.IndexOf(BuildOrderedNodesCache(Count), revision);
+            if (!_loadingCompleted)
+            {
+                // Try to use the existing cache, to avoid rebuilding it multiple times
+                index = Array.IndexOf(BuildOrderedNodesCache(0).ToArray(), revision);
+                if (index >= 0)
+                {
+                    return true;
+                }
+            }
+
+            index = Array.IndexOf(BuildOrderedNodesCache(_revisionCount).ToArray(), revision);
             return index >= 0;
         }
 
         public RevisionGraphRevision? GetNodeForRow(int row)
         {
             // Use a local variable, because the cached list can be reset
-            RevisionGraphRevision[] localOrderedNodesCache = BuildOrderedNodesCache(row);
+            ImmutableArray<RevisionGraphRevision> localOrderedNodesCache = BuildOrderedNodesCache(row);
             if (row >= localOrderedNodesCache.Length)
             {
                 return null;
             }
 
-            return localOrderedNodesCache.ElementAt(row);
+            return localOrderedNodesCache[row];
         }
 
         public IRevisionGraphRow? GetSegmentsForRow(int row)
         {
-            // Use a local variable, because the cached list can be reset
-            IList<RevisionGraphRow> localOrderedRowCache = _orderedRowCache;
-            if (localOrderedRowCache is null || row < 0 || row >= localOrderedRowCache.Count)
+            ImmutableArray<RevisionGraphRevision> localOrderedNodesCache = BuildOrderedNodesCache(row);
+            ImmutableArray<RevisionGraphRow> localOrderedRowCache = _orderedRowCache;
+            if (IsRowCacheDirty(localOrderedRowCache, localOrderedNodesCache) || localOrderedRowCache.Length == 0 || row < 0 || row >= localOrderedRowCache.Length)
```

Comment:
This is safe but `GetSegmentsForRow()` should be called for already cached information only.
If the data is outdated, the graph rendering will be restarted anyway.

Reply:
I am not confident that the thread calling() BuildOrderedRowCache is the same calling GetSegmentsForRow() so I prefer to keep this

Reply:
Yes, in most cases, it will just do a quick check for being up to date.